### ISSUE TITLE
Advice on fine-tuning embedding models 

### ISSUE THREAD

**Issue by zilch42** [MaartenGr/BERTopic#1538]
Hi Maarten, 

I'm starting to look at fine tuning embedding models to make sure they are separating documents on my definition of a topic. I'm only just starting to read about fine tuning now and feeling a bit overwhelmed. Just wondering if you have any experience or advice to push me in the right direction? E.g. LoRA, adapters, `setfit` library, full fine tuning with `model.fit()`? Any guidance on how much training data to use? 

All of the models I'm considering fine tuning come from hugging face and most of them are `sentence_transformers` models. I have triplet data for fine tuning (ie anchor, positive and negative sentences). Can LoRA be used with `sentence_transformers` models? 

Thanks as always.

**Comment by MaartenGr** [comment]
Fine-tuning embedding models can indeed be a bit more involved than one might think beforehand. It depends on the type of data you have (triplets, QA, raw text, etc.), the loss you optimize for, the original model, whether you want to train it from scratch, continue training, or fine-tune the model, etc. 

Having said that, I would advise using the `sentence_transformers` library to fine-tune any embedding model. That can generally also be one that you can find on the [MTEB leaderboard](https://huggingface.co/spaces/mteb/leaderboard). 

LoRA isn't typically used for embedding models for a number of reasons, one of which is that what they optimize over is a bit different. Moreover, as these models are quite a bit smaller than, for example, Llama 2 models, I believe they benefit less from the compressive nature that LoRA aims for. It is a bit more involved than that and you can definitely try it out but I would start with what has shown to work well. 

The guides on the sentence-transformers repo are excellent, but you can start with this one: https://www.sbert.net/docs/training/overview.html#loss-functions

Do note though that with triplets the loss function really matters which one you choose. There is not a one-size fits solution but generally, [MNR Loss](https://www.sbert.net/docs/package_reference/losses.html#multiplenegativesrankingloss) works quite well and has been suggested often by the author of SBERT. It could be different by now but it is definitely a good place to start. 

On another note, the type of negative sentences that you have is **exceedingly** important. If the negative sentences are too easy, then there is little to be learned from the model. Not really a challenge right? It is vital that the negative sentences are actually quite similar to the anchor and positive but without being the right answer, it still needs to be negative. 

In other words, start with the SBERT repo and documentation. Definitely the best place to start.

**Comment by zilch42** [comment]
Thanks Maarten! All sounds like great advice and gives me somewhere to start
